rofl-scheduler/src/state.rs

```diff
@@ -0,0 +1,103 @@
+use std::{collections::BTreeMap, time::Instant};
+
+use base64::{prelude::BASE64_STANDARD, Engine};
+use oasis_runtime_sdk::types::address::Address;
+use oasis_runtime_sdk_rofl_market::types::{Deployment, Instance};
+
+use crate::types::{Permissions, METADATA_KEY_PERMISSIONS};
+
+/// Instance state.
+#[derive(Default, Debug, Clone)]
+pub struct InstanceState {
+    /// Address of the instance administrator.
+    admin: Address,
+    /// Permissions.
+    permissions: Permissions,
+    /// Last deployment.
+    pub last_deployment: Option<Deployment>,
+    /// Last error message corresponding to deploying `last_deployment`.
+    pub last_error: Option<String>,
+    // Whether to ignore instance start until the given time elapses.
+    pub ignore_start_until: Option<Instant>,
+    /// Backoff associated with ignoring instance start.
+    pub ignore_start_backoff: Option<backoff::ExponentialBackoff>,
+}
+
+impl InstanceState {
+    /// Update instance admin.
+    pub fn update_admin(&mut self, instance: &Instance) {
+        self.admin = instance.admin;
+    }
+
+    /// Update configured instance permissions from given deployment.
+    pub fn update_permissions(&mut self, deployment: Option<&Deployment>) {
+        self.permissions = deployment
+            .and_then(|d| Self::parse_permissions(&d.metadata))
+            .unwrap_or_default();
+    }
+
+    /// Check whether a given address has permission to perform the given action.
+    pub fn has_permission(&self, action: &str, address: Address) -> bool {
+        // An administrator always has permission to perform all actions.
+        if self.admin == address {
+            return true;
+        }
+
+        match self.permissions.get(action) {
+            Some(addresses) => addresses.contains(&address),
+            None => false,
+        }
+    }
+
+    /// Parse permissions from deployment metadata.
+    fn parse_permissions(metadata: &BTreeMap<String, String>) -> Option<Permissions> {
+        let permissions = metadata.get(METADATA_KEY_PERMISSIONS)?;
```

Comment:
I prefer having this somewhere else and change signature to `update_permissions(&mut self, permissions: Vec<u8>)` (maybe 2 methods, one for raw permissions and one for normal), as the state shouldn't care how permissions are stored. This way, the test would also simplify.

Reply:
Changed a bit. I don't want to have permission extraction/handling logic in the caller, this was the main reason for moving state into its own module.